src/rules/no_await_in_loop.rs

```diff
@@ -24,12 +21,21 @@ impl LintRule for NoAwaitInLoop {
   }
 
   fn lint_program(
+    &self,
+    _context: &mut Context,
+    _program: &swc_ecmascript::ast::Program,
+  ) {
+    unreachable!();
+  }
+
+  fn lint_program_with_ast_view<'a>(
     &self,
     context: &mut Context,
-    program: &swc_ecmascript::ast::Program,
+    program_info: dprint_swc_ecma_ast_view::ProgramInfo<'a>,
   ) {
-    let mut visitor = NoAwaitInLoopVisitor::new(context);
-    visitor.visit_program(program, program);
+    with_ast_view(program_info, |pg| {
+      NoAwaitInLoopHandler.traverse(pg, context);
+    });
```

Comment:
The "view" should be created really fast, but ideally you would only do this once per file instead of once per rule.

Reply:
Good point. Potentially it could. 
We have to take into account that [`lint_program_with_ast_view` method on `LintRule` trait](https://github.com/denoland/deno_lint/blob/17ce8d0f71b786732f1341fe398bb724b0db109f/src/rules.rs#L100) needs to fall back to `lint_program` if a rule doesn't implement `lint_program_with_ast_view`, because we need to maintain both the existing rule implementations that use swc's AST and new rule implementations that use dprint ast-view's stuff until all the existing rules are replaced.

So, the default implementation of `lint_program_with_ast_view` should look like:

```rust
  /// Executes lint using `dprint-swc-ecma-ast-view`.
  /// Falls back to the `lint_program` method if not implemented.
  fn lint_program_with_ast_view<'a>(
    &self,
    context: &mut Context,
    program: dprint_swc_ecma_ast_view::Program;
,
  ) {
    use dprint_swc_ecma_ast_view::Program::*;
    let swc_program = match program {
      Module(m) => swc_ecmascript::ast::Program::Module(m.inner.clone()),
      Script(s) => swc_ecmascript::ast::Program::Script(s.inner.clone()),
    };
    self.lint_program(context, &swc_program);
  }
```

If the signature of `lint_program_with_ast_view` was like above, we could run all rules like:

```rust
    dprint_swc_ecma_ast_view::with_ast_view(program_info, |pg| {
      // Run builtin rules
      for rule in &self.rules {
        rule.lint_program_with_ast_view(&mut context, pg);
      }
    });
```

This code is fine as the ast-view is created once, but I'm concerned about the cost of cloning swc's Module or Script in these lines:

```rust
      Module(m) => swc_ecmascript::ast::Program::Module(m.inner.clone()),
      Script(s) => swc_ecmascript::ast::Program::Script(s.inner.clone()),
```

Any good ideas?

Reply:
Considering `lint_program` takes a reference to an swc `Program` enum, perhaps instead of cloning, we could introduce a new `Program` enum that instead holds a reference to the swc `Module` or `Script`?

```
pub enum ProgramRef<'a> {
    Module(&'a Module),
    Script(&'a Script),
}
```

Then that could be provided to `lint_program` instead.

Reply:
@dsherret Ahh looks cool! That way we could avoid wasteful cloning. Will rewrite it!